### DISCUSSION TITLE
tile_width and tile_height ignored in tiffsave

### DISCUSSION THREAD

**Discussion by hanspinckaers** [libvips/libvips#2351] [Q&A]
## Bug report

**Describe the bug**
Vips always uses 128x128 tiles, even if argument for tile_width is present.

**To Reproduce**
`vips copy  text.mrxs pyramid.tif[pyramid,tile,tile-width=256,tile-height=256] --vips-progress`

**Expected behavior**
Output like:
`vips temp-5: 272128 x 294144 pixels, 8 threads, 256 x 256 tiles, 256 lines in buffer`

**Actual behavior**
I get:
`vips temp-5: 272128 x 294144 pixels, 8 threads, 128 x 128 tiles, 256 lines in buffer` 

**Environment**
(please complete the following information)
 - OS: Ubuntu 1804
 - Vips: vips-8.11.2

**Additional context**
Maybe the bug is here? Seems like the argument is not used.
https://github.com/libvips/libvips/blob/3732751bf9d22b00ae33d6701e9a823c4d1bea87/libvips/foreign/tiffsave.c#L383

**Comment by jcupitt** [comment]
Hi @HansPinckaers, it should work. I see:

```
$ vips copy k2.jpg x.tif[tile,tile-width=512,tile-height=768]
$ tiffinfo x.tif
TIFF Directory at offset 0xa20008 (10616840)
  Image Width: 1450 Image Length: 2048
  Tile Width: 512 Tile Length: 768
  Resolution: 72.009, 72.009 pixels/inch
  Bits/Sample: 8
  Sample Format: unsigned integer
  Compression Scheme: None
  Photometric Interpretation: RGB color
  Orientation: row 0 top, col 0 lhs
  Samples/Pixel: 3
  Planar Configuration: single image plane
```

The output from `--vips-progress` shows the tiles that libvips is using internally for image rendering, not the ones written to the file.

**Reply by hanspinckaers** [reply]
Ah sorry, thanks!

**Comment by jcupitt** [comment]
The lines you found in `tiffsave.c` set the default values. They are changed when an argument is passed by eg. https://github.com/libvips/libvips/blob/3732751bf9d22b00ae33d6701e9a823c4d1bea87/libvips/foreign/tiffsave.c#L249

**Reply by hanspinckaers** [reply]
Thanks that's helpful when reading the code

**Reply by jcupitt** [reply]
No problem. Those macros build a list of known args on the class, then `vips_object_set_property` handles fetching arguments and updating class members:

https://github.com/libvips/libvips/blob/3732751bf9d22b00ae33d6701e9a823c4d1bea87/libvips/foreign/tiffsave.c#L200

The arg lists are also used to generate the Python / Ruby / C++ / etc. APIs.
